File: services/sessionCleanup.js
/**
 * Session Cleanup Service
 *
 * Periodically cleans up expired sessions from the database
 */

const { deleteExpiredSessions } = require('../db/queries/sessions');
const logger = require('./logger');

/**
 * Run session cleanup
 * @returns {Promise<number>} - Number of sessions deleted (0 if database unavailable)
 */
async function runCleanup() {
    try {
        const deletedCount = await deleteExpiredSessions();
        logger.info({ deletedCount }, 'Session cleanup completed');
        return deletedCount;
    } catch (error) {
        // Check if it's a database connection error
        if (error.message && error.message.includes('Internal server error')) {
            logger.warn('Session cleanup skipped: Database unavailable');
            return 0; // Return 0 instead of throwing - allows server to continue
        }
        
        logger.error({ error: error.message }, 'Session cleanup error');
        // Don't throw - allow cleanup to fail silently so server continues
        return 0;
    }
}

/**
 * Start periodic session cleanup
 * Runs every 6 hours by default
 * @param {number} intervalHours - Cleanup interval in hours (default: 6)
 */
function startPeriodicCleanup(intervalHours = 6) {
    const intervalMs = intervalHours * 60 * 60 * 1000;

    logger.info({ intervalHours }, 'Starting periodic session cleanup');

    // Run cleanup immediately on startup
    runCleanup().catch(err => {
        // Only log if it's not a database connection error (already logged in runCleanup)
        if (!err.message || !err.message.includes('Internal server error')) {
            logger.error({ error: err.message }, 'Initial session cleanup failed');
        }
    });

    // Schedule periodic cleanup
    const intervalId = setInterval(() => {
        runCleanup().catch(err => {
            // Only log if it's not a database connection error (already logged in runCleanup)
            if (!err.message || !err.message.includes('Internal server error')) {
                logger.error({ error: err.message }, 'Periodic session cleanup failed');
            }
        });
    }, intervalMs);

    // Handle graceful shutdown
    process.on('SIGTERM', () => {
        logger.info('Stopping session cleanup service (SIGTERM)');
        clearInterval(intervalId);
    });

    process.on('SIGINT', () => {
        logger.info('Stopping session cleanup service (SIGINT)');
        clearInterval(intervalId);
    });

    return intervalId;
}

module.exports = {
    runCleanup,
    startPeriodicCleanup
};

File: shadow-python/app/services/session_cleanup.py
"""
Session Cleanup Service

Periodically cleans up expired sessions from the database
"""

import asyncio
import signal
from typing import Optional
from app.db.queries.sessions import delete_expired_sessions
from app.services.logger import logger


async def run_cleanup() -> int:
    """
    Run session cleanup
    Returns:
        Number of sessions deleted (0 if database unavailable)
    """
    try:
        deleted_count = await delete_expired_sessions()
        logger.info(f'Session cleanup completed: {deleted_count} sessions deleted', deletedCount=deleted_count)
        return deleted_count
    except Exception as error:
        # Check if it's a database connection error
        error_msg = str(error)
        if 'Internal server error' in error_msg:
            logger.warning('Session cleanup skipped: Database unavailable')
            return 0  # Return 0 instead of throwing - allows server to continue
        
        logger.error(f'Session cleanup error: {error_msg}', error=error_msg)
        # Don't throw - allow cleanup to fail silently so server continues
        return 0


_cleanup_task: Optional[asyncio.Task] = None


def start_periodic_cleanup(interval_hours: int = 6):
    """
    Start periodic session cleanup
    Runs every N hours by default
    Args:
        interval_hours: Cleanup interval in hours (default: 6)
    """
    interval_seconds = interval_hours * 60 * 60

    logger.info(f'Starting periodic session cleanup (every {interval_hours} hours)', intervalHours=interval_hours)

    async def cleanup_loop():
        # Run cleanup immediately on startup
        try:
            await run_cleanup()
        except Exception as err:
            # Only log if it's not a database connection error (already logged in run_cleanup)
            if 'Internal server error' not in str(err):
                logger.error(f'Initial session cleanup failed: {str(err)}', error=str(err))
        
        # Schedule periodic cleanup
        while True:
            await asyncio.sleep(interval_seconds)
            try:
                await run_cleanup()
            except Exception as err:
                # Only log if it's not a database connection error (already logged in run_cleanup)
                if 'Internal server error' not in str(err):
                    logger.error(f'Periodic session cleanup failed: {str(err)}', error=str(err))

    global _cleanup_task
    _cleanup_task = asyncio.create_task(cleanup_loop())

    # Handle graceful shutdown
    def signal_handler(sig, frame):
        logger.info('Stopping session cleanup service')
        if _cleanup_task:
            _cleanup_task.cancel()

    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    return _cleanup_task

